### PR TITLE
feat(gnolang): switch fallthrough

### PR DESCRIPTION
**Pull request by tbruyelle**
Part of #492

Bring support for switch `fallthrough`. This instruction transfers control to the next case, even through this case doesn't match (see https://go.dev/play/p/hXFz6tTwtBI for example).

# Description

When executing a fallthrough, you have to remember what is the current clause index, in order to execute the next one (index+1). For that I used the switch statement attributes, so I created a new name of attribute specifically for that. When a switch clause is executed, the clause index is registered in these attributes. Then, if a fallthrough is met, the attribute is read to determine what is the next clause to execute. There's maybe a better way than using attributes, please say so if it's the case.

The same set of rules as the go-lang has been applied:
- must be in a switch (challenges/switch8.gno)
- must not be in a type switch (challenges/switch9.gno)
- must not be in the last clause (challenges/switch8b.gno)
- must be the last instruction of the clause (challenges/switch8c.gno)

# How has this been tested?

The tests related to gnolang are located in the `tests` folder, to run the ones related to `switch` : 

```sh
$ go test ./tests -v -run `/switch'
```

Note 1: a new `TestChallenges` has been added to cover the files located in `tests/challenges`. Some new files were added (switch6b, switch6c, switch8b, switch8c) to cover other cases.

Note 2: `syncWanted` variable has been set to false (only for the gnolang tests `TestFiles`, `TestFiles2` and the new one `TestChallenges`), so the tests fail if the output doesn't match to the `// Output` comment defined in the code.

All switch cases listed in `tests/challenges` are now working, maybe they should be moved into `tests/files` or `tests/files2` ?

### CODE REVIEW

File: tests/file.go [hunk 1]
```diff
@@ -203,7 +203,7 @@ func RunFileTest(rootDir string, path string, nativeLibs bool, logger loggerFunc
 				// errWanted given
 				if errWanted != "" {
 					if pnc == nil {
-						panic(fmt.Sprintf("fail on %s: got nil error, want: %q", path, errWanted))
+						return fmt.Errorf("fail on %s: got nil error, want: %q", path, errWanted)
```
Comment (tbruyelle) [inline tests/file.go:206]: I found better to return an error here, so the tests are not interrupted.

Comment (tbruyelle) [reply tests/file.go]: Actually break some other tests, so I reverted those.

File: tests/challenges/switch6b.gno [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package main
```
Comment (moul) [inline tests/challenges/switch6b.gno:1]: could you move those challenges to the appropriate folder (outside of `challenges`)?

they'll become official tests and be included in the CI.

**Review by moul — CHANGES_REQUESTED** [review]
Looks great, thank you!

File: tests/challenges/switch6b.gno (hunk 2 above)
Comment (tbruyelle) [reply tests/challenges/switch6b.gno:1]: Should I duplicate those files between `files` and `files2` folder? Or only `files` folder ? 

TBH I don't really get why there's this separation.

Comment (moul) [reply tests/challenges/switch6b.gno:1]: The difference is:
* `files` uses `ImportModeNativePreferred`
* `files2` uses `ImportModeStdlibsPreferred`

See: https://github.com/gnolang/gno/blob/master/tests/imports.go#L61-L63

---

In your case, without any import, `files/` should be enough, but it's cheap to also add it in `files2/` and can prevent eventual future bugs. So, let's do both.

**Review by moul — APPROVED** [review]

File: tests/challenges/switch6b.gno (hunk 2 above)
Comment (moul) [reply tests/challenges/switch6b.gno]: TestChallenges is a good addition, but it should be disabled by default; let's make it toggleable with an env var or something. It's just for advanced developers working on this need.

Goals are:
* CI is green
* running `make test` or `go test ./...` should also work by default

Let's keep the exception just for advanced users.

**Comment by moul** [comment]
@tbruyelle merging now, everything looks good and CI is green.

Feel free to propose changes in makefile, ci, tests etc, in another PR.

Thank you 🙏

**Pull request merged**